### PR TITLE
Set print

### PR DESCRIPTION
**Pull request by srajangarg**
Addresses #10672 

I changed the printing of `FiniteSet` from the bracket-enclosed form `{ , , ... , , }` to `FiniteSet( , , ... , ,)`. This was needed as `{ }` enclosed elements are thought of as python dictionaries by the interpreter, and our requirement is that whatever is output by `str` should be able to reproduce the expression.

Edit: I meant python set, not dictionary.

### CODE REVIEW

**Comment by asmeurer** [comment]
`{1, 2, 3}` represents a set if you use Python 2.7 or higher. We are dropping Python 2.6 support in version 1.0. However, a possible pitfall is that `{}` is an empty dictionary, not the empty set.

**Comment by srajangarg** [comment]
Agreed, but it represents a python set right? and not a Sympy set?

for eg.
```python
>>> Interval(0, 1) * FiniteSet(1, 2)
Interval(0, 1) * FiniteSet(1, 2)
```
But, if we were to call `str` on a `ProductSet` it would produce, `Interval(0, 1) * {1, 2}` which will not be accepted
```python
>>> Interval(0, 1) * {1, 2}
Type Error
```

**Comment by srajangarg** [comment]
Can you tell me what to stick with, so that I can make the modifications in all of the various test files accordingly?

**Comment by srajangarg** [comment]
@asmeurer ping!

**Comment by asmeurer** [comment]
Using `{...}` notation should work with operators after https://github.com/sympy/sympy/pull/10657.

File: sympy/printing/pretty/tests/test_pretty.py [hunk 1]
```diff
@@ -3194,7 +3194,7 @@ def test_pretty_sets():
 
 def test_pretty_ConditionSet():
     from sympy import ConditionSet
-    ascii_str = '{x | x in (-oo, oo) and sin(x) = 0}'
+    ascii_str = '{x | x in Interval(-oo, oo, True, True) and sin(x) = 0}'
```
Comment (asmeurer) [inline sympy/printing/pretty/tests/test_pretty.py]: The pretty printer should still print intervals like `(-oo, oo)`.

**Comment by asmeurer** [comment]
You should make sure that the pretty printers still print things nicely. It is only the string printers that need to be changed.

**Comment by srajangarg** [comment]
Yes, I will double check the pretty printing and all other printers as well.
Even though #10657 is merged, `Interval(0, 1) * {1, 2}` still doesn't work. Am I doing something wrong?

**Comment by srajangarg** [comment]
@asmeurer ping!

**Comment by asmeurer** [comment]
It looks like it doesn't work because ProductSet doesn't call sympify on its arguments.

**Comment by srajangarg** [comment]
Something really weird is happening, or I'm just not able to figure it out.
I added the lines 
```
 print(arg)
print(type(arg))
print(isinstance(arg, Set))
```
in the `flatten` method inside the constructor of `ProductSet` right at the top.
```
>>> Interval(0, 1) * {1, 2}
[[0, 1], set([1, 2])]
<type 'list'>
False
[0, 1]
<class 'sympy.sets.sets.Interval'>
False
Traceback ----
TypeError: Input must be Sets or iterables of Sets
```
What's weird is that when `arg` is `[0,1]` of type `Interval`, `isinstance(Set)` is returning `False`!
Is this expected behavior?

**Comment by srajangarg** [comment]
@asmeurer

**Comment by srajangarg** [comment]
Also, I don't seem to have changed anything in `printing/pretty.py`, so why is pretty printing affected? (as you pointed earlier) Help is really appreciated.

**Comment by asmeurer** [comment]
I don't know what's causing that. I can't reproduce it myself. For me, this patch fixes the issue

```patch
diff --git a/sympy/sets/sets.py b/sympy/sets/sets.py
index 299fd1d..f1865bc 100644
--- a/sympy/sets/sets.py
+++ b/sympy/sets/sets.py
@@ -556,6 +556,7 @@ class ProductSet(Set):
     is_ProductSet = True

     def __new__(cls, *sets, **assumptions):
+        sets = _sympify(sets)
         def flatten(arg):
             if isinstance(arg, Set):
                 if arg.is_ProductSet:
```

**Comment by srajangarg** [comment]
Yes, your patch worked. But some problems will still remain. What do you suggest we do about this?
```python
ps = FiniteSet(0, 1) * FiniteSet(1, 2)
>>> str(ps)
{0, 1} * {1, 2}
>>> {0, 1} * {1, 2}
TypeError: unsupported operand type(s) for *: 'set' and 'set'
```
Also any clue about the pretty printing problem?

**Comment by srajangarg** [comment]
@asmeurer  I'm helpless! Intervals are behaving really weird. (current PR)
```python
>>> a = ConditionSet(x, Eq(sin(x), 0), S.Reals)
>>> r = a.base_set
Interval(-oo, oo, True, True)
>>> l = Interval(-oo, oo, True, True)
Interval(-oo, oo, True, True)
>>> pprint(l, use_unicode=False)
(-oo, oo)
>>> pprint(r, use_unicode=False)
Interval(-oo, oo, True, True)
```
Evetn though `l` and `r` are exactly the same!

I looked at the trace, and it's somehow reaching `printing/pretty/pretty_symology.py:107`, where `str()` is called and hence it's not printing pretty! Only in the case of `r` and not `l` !!
```python
def xstr(*args):
    """call str or unicode depending on current mode"""
    if _use_unicode:
        return unicode(*args)
    else:
        return str(*args)
```

**Comment by asmeurer** [comment]
There's no way to make `{0, 1}*{0, 1}` work. There are no SymPy types involved in that operation. It's all Python. Python doesn't let you change what operators between two builtin types do. So the printer will need to be smart enough to not use `{}` in that case.

**Comment by asmeurer** [comment]
For the interval thing, it looks weird indeed. I recommend getting a debugger and stepping through the code for each pprint, and seeing where and why they diverge. I personally use PuDB, but there are many other good debuggers as well.

**Pull request closed without merging**